[Java] Support jdk record serialization

## Is your feature request related to a problem? Please describe.
Record is final in JDK16, the record is a final object.

Jdk has a different object model for it, we can't get field offset for it, thus can't use reflection/methodhandle to set its fields. Instead, we need to call its constructor with all its friends to recreate the object.
```
Caused by: java.lang.UnsupportedOperationException: can't get field offset on a record class: private final int org.example.TestRecord$Foo.f1
	at jdk.unsupported/sun.misc.Unsafe.objectFieldOffset(Unsafe.java:648)
```
![image](https://github.com/alipay/fury/assets/12445254/f0258319-a84d-45f6-8788-c2b7770baa22)

The record class can envolve too, and jdk support record class fields forward/backward compatibility, fury should support too.

Another thing to notice is that record class doesn't allow customize serialization, which makes serialization simple. since we don't have to support `writeObject/readObject/writeReplace/readResolve/readObjectNoData`:
- Serialization of a record object is based only on its state components.
- Deserialization of a record object uses only the canonical constructor.


## Describe the solution you'd like
- Make serialization use record class accessor method to get component value
- Read record class fields first, then use record class constructor to recreate record.

Steps:
- [x] #838
- [x] #839
- [x] #842
- [x] #840
- [x] #841
- [x] #843


## Additional context
- https://www.infoq.com/articles/java-14-feature-spotlight/
- https://inside.java/2020/07/20/record-serialization/
- https://blogs.oracle.com/javamagazine/post/simpler-object-and-data-serialization-using-java-records
- #828

--- a/integration_tests/latest_jdk_tests/src/test/java/io/fury/integration_tests/RecordSerializersTest.java
+++ b/integration_tests/latest_jdk_tests/src/test/java/io/fury/integration_tests/RecordSerializersTest.java
@@ -129,8 +129,8 @@ public class RecordSerializersTest {
     Assert.assertEquals(fury2.deserialize(bytes1), record2);
   }
 
-  @Test
-  public void testRecordMetaShare() throws Throwable {
+  @Test(dataProvider = "codegen")
+  public void testRecordMetaShare(boolean codegen) throws Throwable {
     String code1 =
         "import java.util.*;"
             + "public record TestRecord(int f1, String f2, List<String> f3, char f4, Map<String, Integer> f5) {}";
@@ -142,7 +142,7 @@ public class RecordSerializersTest {
     Fury fury1 =
         Fury.builder()
             .requireClassRegistration(false)
-            .withCodegen(false)
+            .withCodegen(codegen)
             .withCompatibleMode(CompatibleMode.COMPATIBLE)
             .withMetaContextShare(true)
             .withClassLoader(cls1.getClassLoader())
@@ -156,7 +156,7 @@ public class RecordSerializersTest {
     Fury fury2 =
         Fury.builder()
             .requireClassRegistration(false)
-            .withCodegen(false)
+            .withCodegen(codegen)
             .withCompatibleMode(CompatibleMode.COMPATIBLE)
             .withMetaContextShare(true)
             .withClassLoader(cls2.getClassLoader())
@@ -172,5 +172,43 @@ public class RecordSerializersTest {
     fury1.getSerializationContext().setMetaContext(metaContext1);
     Object o12 = fury1.deserialize(bytes2);
     System.out.println(o12);
+  }
+
+  @Test(dataProvider = "codegen")
+  public void testPrivateRecords(boolean codegen) {
+    {
+      Fury fury = Fury.builder().requireClassRegistration(false).withCodegen(codegen).build();
+      Object o1 = Records.createPrivateRecord(11);
+      Assert.assertEquals(fury.deserialize(fury.serialize(o1)), o1);
+      Object o2 = Records.createPublicRecord(11, o1);
+      Assert.assertEquals(fury.deserialize(fury.serialize(o2)), o2);
+    }
+    {
+      Fury fury =
+          Fury.builder()
+              .requireClassRegistration(false)
+              .withCodegen(codegen)
+              .withCompatibleMode(CompatibleMode.COMPATIBLE)
+              .build();
+      Object o1 = Records.createPrivateRecord(11);
+      Assert.assertEquals(fury.deserialize(fury.serialize(o1)), o1);
+      Object o2 = Records.createPublicRecord(11, o1);
+      Assert.assertEquals(fury.deserialize(fury.serialize(o2)), o2);
+    }
+    {
+      Fury fury =
+          Fury.builder()
+              .requireClassRegistration(false)
+              .withCodegen(codegen)
+              .withMetaContextShare(true)
+              .build();
+      Object o1 = Records.createPrivateRecord(11);
+      Object o2 = Records.createPublicRecord(11, o1);
+      MetaContext context = new MetaContext();
+      fury.getSerializationContext().setMetaContext(context);
+      byte[] bytes = fury.serialize(o2);
+      fury.getSerializationContext().setMetaContext(context);
+      Assert.assertEquals(fury.deserialize(bytes), o2);
+    }
   }
 }

--- a/integration_tests/latest_jdk_tests/src/test/java/io/fury/integration_tests/Records.java
+++ b/integration_tests/latest_jdk_tests/src/test/java/io/fury/integration_tests/Records.java
@@ -1,0 +1,15 @@
+package io.fury.integration_tests;
+
+public class Records {
+  private record PrivateRecord(int f1) {}
+
+  public record PublicRecord(int f1, PrivateRecord f2) {}
+
+  public static Object createPrivateRecord(int f1) {
+    return new PrivateRecord(f1);
+  }
+
+  public static Object createPublicRecord(int f1, Object privateObject) {
+    return new PublicRecord(f1, (PrivateRecord) privateObject);
+  }
+}

--- a/integration_tests/latest_jdk_tests/src/test/java/io/fury/integration_tests/Records.java
+++ b/integration_tests/latest_jdk_tests/src/test/java/io/fury/integration_tests/Records.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2023 The Fury Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package io.fury.integration_tests;
 
 public class Records {

--- a/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
@@ -439,7 +439,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
               "loadClass",
               CLASS_TYPE,
               beanClassExpr(),
-              Literal.ofString(cls.getSimpleName() + ".class"));
+              Literal.ofString(cls.getName()));
       // Don't invoke `Serializer.newSerializer` here, since it(ex. ObjectSerializer) may set itself
       // as global serializer, which overwrite serializer updates in jit callback.
       Expression newSerializerExpr =

--- a/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
@@ -52,6 +52,7 @@ import io.fury.codegen.Expression.Cast;
 import io.fury.codegen.Expression.If;
 import io.fury.codegen.Expression.Invoke;
 import io.fury.codegen.Expression.ListExpression;
+import io.fury.codegen.Expression.Literal;
 import io.fury.codegen.Expression.Reference;
 import io.fury.codegen.Expression.Return;
 import io.fury.codegen.ExpressionOptimizer;
@@ -271,14 +272,11 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
       Expression action =
           new ListExpression(
               new Invoke(
-                  buffer,
-                  "writeByte",
-                  new Expression.Literal(Fury.REF_VALUE_FLAG, PRIMITIVE_BYTE_TYPE)),
+                  buffer, "writeByte", new Literal(Fury.REF_VALUE_FLAG, PRIMITIVE_BYTE_TYPE)),
               serializeForNotNull(inputObject, buffer, typeToken, generateNewMethod));
       return new If(
           ExpressionUtils.eqNull(inputObject),
-          new Invoke(
-              buffer, "writeByte", new Expression.Literal(Fury.NULL_FLAG, PRIMITIVE_BYTE_TYPE)),
+          new Invoke(buffer, "writeByte", new Literal(Fury.NULL_FLAG, PRIMITIVE_BYTE_TYPE)),
           action);
     }
   }
@@ -435,11 +433,17 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
         serializerClass = Serializer.class;
       }
       TypeToken<? extends Serializer> serializerTypeToken = TypeToken.of(serializerClass);
-      Expression.Literal clzLiteral = new Expression.Literal(ctx.type(cls) + ".class");
+      Expression.StaticInvoke fieldTypeExpr =
+          new Expression.StaticInvoke(
+              ReflectionUtils.class,
+              "loadClass",
+              CLASS_TYPE,
+              beanClassExpr(),
+              Literal.ofString(cls.getSimpleName() + ".class"));
       // Don't invoke `Serializer.newSerializer` here, since it(ex. ObjectSerializer) may set itself
       // as global serializer, which overwrite serializer updates in jit callback.
       Expression newSerializerExpr =
-          inlineInvoke(classResolverRef, "getSerializer", SERIALIZER_TYPE, clzLiteral);
+          inlineInvoke(classResolverRef, "getSerializer", SERIALIZER_TYPE, fieldTypeExpr);
       String name = ctx.newName(StringUtils.uncapitalize(serializerClass.getSimpleName()));
       // It's ok it jit already finished and this method return false, in such cases
       // `serializerClass` is already jit generated class.
@@ -475,7 +479,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
       if (classInfoRef != null) {
         return Tuple2.of(classInfoRef, false);
       }
-      Expression clsExpr = new Expression.Literal(cls, CLASS_TYPE);
+      Expression clsExpr = new Literal(cls, CLASS_TYPE);
       classInfoExpr = inlineInvoke(classResolverRef, "getClassInfo", classInfoTypeToken, clsExpr);
       // Use `ctx.freshName(cls)` to avoid wrong name for arr type.
       String name = ctx.newName(ctx.newName(cls) + "ClassInfo");
@@ -544,7 +548,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
     Expression clsExpr;
     if (isFinal(clz)) {
       serializer = getOrCreateSerializer(clz);
-      clsExpr = new Expression.Literal(clz, CLASS_TYPE);
+      clsExpr = new Literal(clz, CLASS_TYPE);
     } else {
       ListExpression writeClassAction = new ListExpression();
       Tuple2<Reference, Boolean> classInfoRef = addClassInfoField(clz);
@@ -576,7 +580,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
       // if type hierarchy no interact, cast will compile error.
       hookWrite =
           new If(
-              eq(clsExpr, new Expression.Literal(ArrayList.class, CLASS_TYPE)),
+              eq(clsExpr, new Literal(ArrayList.class, CLASS_TYPE)),
               writeListCodegen(buffer, collection, serializer, elementType),
               writeCollectionCodegen(buffer, collection, serializer, elementType),
               false);
@@ -617,8 +621,8 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
       Expression buffer, Expression collection, Expression serializer, TypeToken<?> elementType) {
     Expression list =
         new Cast(collection, TypeUtils.arrayListOf(getRawType(elementType)), "arrList");
-    Expression start = new Expression.Literal(0, PRIMITIVE_INT_TYPE);
-    Expression step = new Expression.Literal(1, PRIMITIVE_INT_TYPE);
+    Expression start = new Literal(0, PRIMITIVE_INT_TYPE);
+    Expression step = new Literal(1, PRIMITIVE_INT_TYPE);
     Invoke size = new Invoke(collection, "size", PRIMITIVE_INT_TYPE);
     Invoke writeSize = new Invoke(buffer, "writePositiveVarInt", size);
     Invoke writeHeader = new Invoke(serializer, "writeHeader", buffer, collection);
@@ -779,8 +783,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
       Expression refId = tryPreserveRefId(buffer);
       // indicates that the object is first read.
       Expression needDeserialize =
-          ExpressionUtils.egt(
-              refId, new Expression.Literal(Fury.NOT_NULL_VALUE_FLAG, PRIMITIVE_BYTE_TYPE));
+          ExpressionUtils.egt(refId, new Literal(Fury.NOT_NULL_VALUE_FLAG, PRIMITIVE_BYTE_TYPE));
       Expression deserializedValue = deserializeForNotNull(buffer, typeToken, generateNewMethod);
       Expression setReadObject =
           new Invoke(refResolverRef, "setReadObject", refId, deserializedValue);
@@ -801,7 +804,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
       Expression notNull =
           neq(
               inlineInvoke(buffer, "readByte", PRIMITIVE_BYTE_TYPE),
-              new Expression.Literal(Fury.NULL_FLAG, PRIMITIVE_BYTE_TYPE));
+              new Literal(Fury.NULL_FLAG, PRIMITIVE_BYTE_TYPE));
       Expression value = deserializeForNotNull(buffer, typeToken, generateNewMethod);
       // use false to ignore null.
       return new If(
@@ -913,8 +916,8 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
 
   protected Expression readCollectionCodegen(
       Expression buffer, Expression collection, Expression size, TypeToken<?> elementType) {
-    Expression start = new Expression.Literal(0, PRIMITIVE_INT_TYPE);
-    Expression step = new Expression.Literal(1, PRIMITIVE_INT_TYPE);
+    Expression start = new Literal(0, PRIMITIVE_INT_TYPE);
+    Expression step = new Literal(1, PRIMITIVE_INT_TYPE);
     ExpressionVisitor.ExprHolder exprHolder =
         ExpressionVisitor.ExprHolder.of("collection", collection, "buffer", buffer);
     Expression.ForLoop readElements =
@@ -957,8 +960,8 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
     Invoke supportHook = inlineInvoke(serializer, "supportCodegenHook", PRIMITIVE_BOOLEAN_TYPE);
     Expression size = new Invoke(buffer, "readPositiveVarInt", "size", PRIMITIVE_INT_TYPE);
     Expression newMap = new Invoke(serializer, "newMap", MAP_TYPE, buffer, size);
-    Expression start = new Expression.Literal(0, PRIMITIVE_INT_TYPE);
-    Expression step = new Expression.Literal(1, PRIMITIVE_INT_TYPE);
+    Expression start = new Literal(0, PRIMITIVE_INT_TYPE);
+    Expression step = new Literal(1, PRIMITIVE_INT_TYPE);
     ExpressionVisitor.ExprHolder exprHolder =
         ExpressionVisitor.ExprHolder.of("map", newMap, "buffer", buffer);
     Expression.ForLoop readKeyValues =

--- a/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
@@ -55,6 +55,7 @@ import io.fury.codegen.Expression.ListExpression;
 import io.fury.codegen.Expression.Literal;
 import io.fury.codegen.Expression.Reference;
 import io.fury.codegen.Expression.Return;
+import io.fury.codegen.Expression.StaticInvoke;
 import io.fury.codegen.ExpressionOptimizer;
 import io.fury.codegen.ExpressionUtils;
 import io.fury.codegen.ExpressionVisitor;
@@ -433,13 +434,18 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
         serializerClass = Serializer.class;
       }
       TypeToken<? extends Serializer> serializerTypeToken = TypeToken.of(serializerClass);
-      Expression.StaticInvoke fieldTypeExpr =
-          new Expression.StaticInvoke(
-              ReflectionUtils.class,
-              "loadClass",
-              CLASS_TYPE,
-              beanClassExpr(),
-              Literal.ofString(cls.getName()));
+      Expression fieldTypeExpr;
+      if (Modifier.isPublic(cls.getModifiers())) {
+        fieldTypeExpr = Literal.ofClass(cls);
+      } else {
+        fieldTypeExpr =
+            new StaticInvoke(
+                ReflectionUtils.class,
+                "loadClass",
+                CLASS_TYPE,
+                beanClassExpr(),
+                Literal.ofString(cls.getName()));
+      }
       // Don't invoke `Serializer.newSerializer` here, since it(ex. ObjectSerializer) may set itself
       // as global serializer, which overwrite serializer updates in jit callback.
       Expression newSerializerExpr =

--- a/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
@@ -250,7 +250,8 @@ public abstract class CodecBuilder {
               fieldNullable,
               inputObject,
               fieldOffsetExpr);
-      return new Expression.Cast(getObj, descriptor.getTypeToken(), fieldName);
+      TypeToken<?> publicSuperType = ReflectionUtils.getPublicSuperType(descriptor.getTypeToken());
+      return new Expression.Cast(getObj, publicSuperType, fieldName);
     }
   }
 

--- a/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
@@ -39,7 +39,6 @@ import io.fury.codegen.Expression.Invoke;
 import io.fury.codegen.Expression.Literal;
 import io.fury.codegen.Expression.Reference;
 import io.fury.codegen.Expression.StaticInvoke;
-import io.fury.codegen.ExpressionUtils;
 import io.fury.collection.Tuple2;
 import io.fury.memory.MemoryBuffer;
 import io.fury.resolver.ClassInfo;
@@ -329,7 +328,7 @@ public abstract class CodecBuilder {
       TypeToken<Field> fieldTypeToken = TypeToken.of(Field.class);
       String fieldRefName = ctx.newName(fieldName + "Field");
       Preconditions.checkArgument(Modifier.isPublic(cls.getModifiers()));
-      Literal clzLiteral = new Literal(ctx.type(cls) + ".class");
+      Literal clzLiteral = Literal.ofClass(cls);
       StaticInvoke fieldExpr =
           new StaticInvoke(
               ReflectionUtils.class,
@@ -337,8 +336,8 @@ public abstract class CodecBuilder {
               fieldTypeToken,
               false,
               clzLiteral,
-              ExpressionUtils.literalStr(fieldName));
-      Invoke setAccessible = new Invoke(fieldExpr, "setAccessible", new Literal("true"));
+              Literal.ofString(fieldName));
+      Invoke setAccessible = new Invoke(fieldExpr, "setAccessible", Literal.True);
       Expression.ListExpression createField =
           new Expression.ListExpression(setAccessible, fieldExpr);
       ctx.addField(ctx.type(Field.class), fieldRefName, createField);

--- a/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
@@ -35,10 +35,12 @@ import io.fury.Fury;
 import io.fury.codegen.CodegenContext;
 import io.fury.codegen.Expression;
 import io.fury.codegen.Expression.Inlineable;
+import io.fury.codegen.Expression.Invoke;
 import io.fury.codegen.Expression.Literal;
 import io.fury.codegen.Expression.Reference;
 import io.fury.codegen.Expression.StaticInvoke;
 import io.fury.codegen.ExpressionUtils;
+import io.fury.collection.Tuple2;
 import io.fury.memory.MemoryBuffer;
 import io.fury.resolver.ClassInfo;
 import io.fury.resolver.ClassInfoCache;
@@ -142,10 +144,9 @@ public abstract class CodecBuilder {
   /** Returns an expression that get field value from <code>bean</code>. */
   protected Expression getFieldValue(Expression inputBeanExpr, Descriptor descriptor) {
     TypeToken<?> fieldType = descriptor.getTypeToken();
-    // No public field type is cast to public parent classes in DescriptorGrouper#createDescriptor
-    Preconditions.checkArgument(
-        Modifier.isPublic(getRawType(fieldType).getModifiers()),
-        "Field type should be public for codegen-based access");
+    if (!Modifier.isPublic(getRawType(fieldType).getModifiers())) {
+      fieldType = OBJECT_TYPE;
+    }
     String fieldName = descriptor.getName();
     if (isRecord) {
       return getRecordFieldValue(inputBeanExpr, descriptor);
@@ -158,7 +159,7 @@ public abstract class CodecBuilder {
       return new Expression.FieldValue(inputBeanExpr, fieldName, fieldType, fieldNullable, false);
     } else if (descriptor.getReadMethod() != null
         && Modifier.isPublic(descriptor.getReadMethod().getModifiers())) {
-      return new Expression.Invoke(
+      return new Invoke(
           inputBeanExpr, descriptor.getReadMethod().getName(), fieldName, fieldType, fieldNullable);
     } else {
       if (!Modifier.isPrivate(descriptor.getModifiers())) {
@@ -191,22 +192,28 @@ public abstract class CodecBuilder {
     String fieldName = descriptor.getName();
     if (Modifier.isPublic(beanClass.getModifiers())) {
       Preconditions.checkNotNull(descriptor.getReadMethod());
-      return new Expression.Invoke(
+      return new Invoke(
           inputBeanExpr, descriptor.getReadMethod().getName(), fieldName, fieldType, fieldNullable);
     } else {
       String key = "_" + fieldName + "_getter_";
       Reference ref = fieldMap.get(key);
+      Tuple2<Class<?>, String> methodInfo = Functions.getterMethodInfo(descriptor.getRawType());
       if (ref == null) {
-        Object getterFunction = Functions.makeGetterFunction(beanClass, fieldName);
-        TypeToken<?> getterType = TypeToken.of(getterFunction.getClass());
-        StaticInvoke getter =
+        Class<?> funcInterface = methodInfo.f0;
+        TypeToken<?> getterType = TypeToken.of(funcInterface);
+        Expression getter =
             new StaticInvoke(
-                Functions.class, "makeGetterFunction", getterType, Literal.ofString(fieldName));
-        ctx.addField(getterFunction.getClass(), key, getter);
-        ref = new Reference(fieldName, getterType);
-        fieldMap.put(fieldName, ref);
+                Functions.class,
+                "makeGetterFunction",
+                OBJECT_TYPE,
+                beanClassExpr(),
+                Literal.ofString(fieldName));
+        getter = new Expression.Cast(getter, getterType);
+        ctx.addField(funcInterface, key, getter);
+        ref = new Reference(key, getterType);
+        fieldMap.put(key, ref);
       }
-      return ref;
+      return new Invoke(ref, methodInfo.f1, fieldType, fieldNullable, inputBeanExpr);
     }
   }
 
@@ -215,8 +222,7 @@ public abstract class CodecBuilder {
       Expression inputObject, Class<?> cls, Descriptor descriptor) {
     Reference fieldRef = getOrCreateField(cls, descriptor.getName());
     // boolean fieldNullable = !descriptor.getTypeToken().isPrimitive();
-    Expression.Invoke getObj =
-        new Expression.Invoke(fieldRef, "get", OBJECT_TYPE, fieldNullable, inputObject);
+    Invoke getObj = new Invoke(fieldRef, "get", OBJECT_TYPE, fieldNullable, inputObject);
     return new Expression.Cast(getObj, descriptor.getTypeToken(), descriptor.getName());
   }
 
@@ -267,7 +273,7 @@ public abstract class CodecBuilder {
     if (!Modifier.isFinal(d.getModifiers()) && Modifier.isPublic(d.getModifiers())) {
       return new Expression.SetField(bean, fieldName, value);
     } else if (d.getWriteMethod() != null && Modifier.isPublic(d.getWriteMethod().getModifiers())) {
-      return new Expression.Invoke(bean, d.getWriteMethod().getName(), value);
+      return new Invoke(bean, d.getWriteMethod().getName(), value);
     } else {
       if (!Modifier.isFinal(d.getModifiers()) && !Modifier.isPrivate(d.getModifiers())) {
         if (AccessorHelper.defineAccessor(d.getField())) {
@@ -295,7 +301,7 @@ public abstract class CodecBuilder {
     // populate fieldMap
     Reference fieldRef = getOrCreateField(getRawType(bean.type()), fieldName);
     Preconditions.checkNotNull(fieldRef);
-    return new Expression.Invoke(fieldRef, "set", bean, value);
+    return new Invoke(fieldRef, "set", bean, value);
   }
 
   /**
@@ -332,8 +338,7 @@ public abstract class CodecBuilder {
               false,
               clzLiteral,
               ExpressionUtils.literalStr(fieldName));
-      Expression.Invoke setAccessible =
-          new Expression.Invoke(fieldExpr, "setAccessible", new Literal("true"));
+      Invoke setAccessible = new Invoke(fieldExpr, "setAccessible", new Literal("true"));
       Expression.ListExpression createField =
           new Expression.ListExpression(setAccessible, fieldExpr);
       ctx.addField(ctx.type(Field.class), fieldRefName, createField);

--- a/java/fury-core/src/main/java/io/fury/codegen/Expression.java
+++ b/java/fury-core/src/main/java/io/fury/codegen/Expression.java
@@ -2092,7 +2092,7 @@ public interface Expression {
       String i = ctx.newName("i");
       String elemValue = ctx.newName("elemValue");
       Expression elementExpr =
-          action.apply(new Literal(i), new Reference(elemValue, elementType, false));
+          action.apply(new Reference(i), new Reference(elemValue, elementType, false));
       ExprCode elementExprCode = elementExpr.genCode(ctx);
 
       if (inputObject.type().isArray()) {
@@ -2227,7 +2227,7 @@ public interface Expression {
       rightElemType = ReflectionUtils.getPublicSuperType(rightElemType);
       Expression elemExpr =
           action.apply(
-              new Literal(i),
+              new Reference(i),
               new Reference(leftElemValue, leftElemType, true),
               // elemValue nullability check use isNullAt inside action, so elemValueRef'nullable is
               // false.

--- a/java/fury-core/src/main/java/io/fury/codegen/Expression.java
+++ b/java/fury-core/src/main/java/io/fury/codegen/Expression.java
@@ -283,7 +283,7 @@ public interface Expression {
         return new ExprCode(null, TrueLiteral, defaultLiteral);
       } else {
         if (javaType == String.class) {
-          return new ExprCode(FalseLiteral, new LiteralValue(value));
+          return new ExprCode(FalseLiteral, new LiteralValue("\"" + value + "\""));
         } else if (javaType == Boolean.class || javaType == Integer.class) {
           return new ExprCode(null, FalseLiteral, new LiteralValue(javaType, value.toString()));
         } else if (javaType == Float.class) {

--- a/java/fury-core/src/main/java/io/fury/codegen/ExpressionUtils.java
+++ b/java/fury-core/src/main/java/io/fury/codegen/ExpressionUtils.java
@@ -74,11 +74,6 @@ public class ExpressionUtils {
     return new Literal(null, type);
   }
 
-  public static Literal literalStr(String value) {
-    value = String.format("\"%s\"", value);
-    return new Literal(value);
-  }
-
   public static Comparator eq(Expression left, Expression right) {
     return new Comparator("==", left, right, true);
   }

--- a/java/fury-core/src/main/java/io/fury/type/DescriptorGrouper.java
+++ b/java/fury-core/src/main/java/io/fury/type/DescriptorGrouper.java
@@ -20,8 +20,6 @@ import static io.fury.type.TypeUtils.getSizeOfPrimitiveType;
 
 import com.google.common.collect.ImmutableSet;
 import com.google.common.primitives.Primitives;
-import com.google.common.reflect.TypeToken;
-import io.fury.util.ReflectionUtils;
 import io.fury.util.record.RecordUtils;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;

--- a/java/fury-core/src/main/java/io/fury/type/DescriptorGrouper.java
+++ b/java/fury-core/src/main/java/io/fury/type/DescriptorGrouper.java
@@ -197,20 +197,11 @@ public class DescriptorGrouper {
     if (readMethod != null && !RecordUtils.isRecord(readMethod.getDeclaringClass())) {
       readMethod = null;
     }
-    if (!Modifier.isPublic(d.getRawType().getModifiers())) {
-      // Non-public class can't be accessed from generated code.
-      // (Ignore protected/package level access for simplicity.
-      // Since class members whose type are non-public class are rare,
-      // it doesn't have much impact on performance.)
-      TypeToken<?> publicSuperType = ReflectionUtils.getPublicSuperType(d.getTypeToken());
-      return d.copy(publicSuperType, readMethod, null);
-    } else {
-      // getter/setter may lose some inner state of an object, so we set them to null.
-      if (readMethod == null && d.getWriteMethod() == null) {
-        return d;
-      }
-      return d.copy(d.getTypeToken(), readMethod, null);
+    // getter/setter may lose some inner state of an object, so we set them to null.
+    if (readMethod == null && d.getWriteMethod() == null) {
+      return d;
     }
+    return d.copy(d.getTypeToken(), readMethod, null);
   }
 
   public static DescriptorGrouper createDescriptorGrouper(

--- a/java/fury-core/src/main/java/io/fury/util/ReflectionUtils.java
+++ b/java/fury-core/src/main/java/io/fury/util/ReflectionUtils.java
@@ -398,6 +398,15 @@ public class ReflectionUtils {
     return pkg;
   }
 
+  // Invoked by JIT.
+  public Class<?> loadClass(Class<?> neighbor, String className) {
+    try {
+      return neighbor.getClassLoader().loadClass(className);
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
   public static <T> T unsafeCopy(T obj) {
     @SuppressWarnings("unchecked")
     T newInstance = (T) Platform.newInstance(obj.getClass());

--- a/java/fury-core/src/main/java/io/fury/util/ReflectionUtils.java
+++ b/java/fury-core/src/main/java/io/fury/util/ReflectionUtils.java
@@ -399,7 +399,7 @@ public class ReflectionUtils {
   }
 
   // Invoked by JIT.
-  public Class<?> loadClass(Class<?> neighbor, String className) {
+  public static Class<?> loadClass(Class<?> neighbor, String className) {
     try {
       return neighbor.getClassLoader().loadClass(className);
     } catch (ClassNotFoundException e) {

--- a/java/fury-core/src/main/java/io/fury/util/function/Functions.java
+++ b/java/fury-core/src/main/java/io/fury/util/function/Functions.java
@@ -17,6 +17,7 @@
 package io.fury.util.function;
 
 import com.google.common.base.Preconditions;
+import io.fury.collection.Tuple2;
 import io.fury.util.ReflectionUtils;
 import io.fury.util.unsafe._JDKAccess;
 import java.io.Serializable;
@@ -84,5 +85,9 @@ public class Functions {
     } catch (IllegalAccessException ex) {
       throw new RuntimeException(ex);
     }
+  }
+
+  public static Tuple2<Class<?>, String> getterMethodInfo(Class<?> type) {
+    return _JDKAccess.getterMethodInfo(type);
   }
 }

--- a/java/fury-core/src/main/java/io/fury/util/unsafe/_JDKAccess.java
+++ b/java/fury-core/src/main/java/io/fury/util/unsafe/_JDKAccess.java
@@ -260,6 +260,10 @@ public class _JDKAccess {
     methodMap.put(double.class, Tuple2.of(ToDoubleFunction.class, "applyAsDouble"));
   }
 
+  public static Tuple2<Class<?>, String> getterMethodInfo(Class<?> type) {
+    return methodMap.get(type);
+  }
+
   public static Object makeGetterFunction(
       MethodHandles.Lookup lookup, MethodHandle handle, Class<?> returnType) {
     Tuple2<Class<?>, String> methodInfo = methodMap.get(returnType);

--- a/java/fury-core/src/test/java/io/fury/builder/ObjectCodecBuilderTest.java
+++ b/java/fury-core/src/test/java/io/fury/builder/ObjectCodecBuilderTest.java
@@ -25,6 +25,7 @@ import io.fury.Language;
 import io.fury.codegen.CodeGenerator;
 import io.fury.codegen.CompileUnit;
 import io.fury.codegen.JaninoUtils;
+import io.fury.test.bean.AccessBeans;
 import io.fury.test.bean.BeanA;
 import io.fury.test.bean.BeanB;
 import io.fury.test.bean.Foo;
@@ -159,5 +160,12 @@ public class ObjectCodecBuilderTest extends FuryTestBase {
     nestedContainer.map1 = map1;
     serDeCheck(fury, nestedContainer);
     checkMethodSize(NestedContainer.class, fury);
+  }
+
+  @Test
+  public void testAccessLevel() {
+    Fury fury = Fury.builder().requireClassRegistration(false).build();
+    AccessBeans.PublicClass object = AccessBeans.createPublicClassObject();
+    serDeCheckSerializer(fury, object, "Codec");
   }
 }

--- a/java/fury-core/src/test/java/io/fury/codegen/ExpressionVisitorTest.java
+++ b/java/fury-core/src/test/java/io/fury/codegen/ExpressionVisitorTest.java
@@ -20,6 +20,7 @@ import static org.testng.Assert.assertEquals;
 
 import com.google.common.base.Preconditions;
 import com.google.common.reflect.TypeToken;
+import io.fury.codegen.Expression.Literal;
 import io.fury.util.ReflectionUtils;
 import java.lang.invoke.SerializedLambda;
 import java.lang.reflect.InvocationTargetException;
@@ -36,9 +37,9 @@ public class ExpressionVisitorTest {
     Expression.Reference ref =
         new Expression.Reference("a", TypeToken.of(ExpressionVisitorTest.class));
     Expression e1 = new Expression.Invoke(ref, "testTraverseExpression");
-    Expression.Literal start = new Expression.Literal("0");
-    Expression.Literal end = new Expression.Literal("10");
-    Expression.Literal step = new Expression.Literal("1");
+    Literal start = Literal.ofInt(0);
+    Literal end = Literal.ofInt(10);
+    Literal step = Literal.ofInt(1);
     ExpressionVisitor.ExprHolder holder =
         ExpressionVisitor.ExprHolder.of("e1", e1, "e2", new Expression.ListExpression());
     // FIXME ListExpression#add in lambda don't get executed, so ListExpression is the last expr.

--- a/java/fury-format/src/main/java/io/fury/format/encoder/ArrayDataForEach.java
+++ b/java/fury-format/src/main/java/io/fury/format/encoder/ArrayDataForEach.java
@@ -97,7 +97,7 @@ public class ArrayDataForEach implements Expression {
     // elemValue is only used in notNullAction, so set elemValueRef'nullable to false.
     Reference elemValueRef = new Reference(elemValue, elemType);
     Code.ExprCode notNullElemExprCode =
-        notNullAction.apply(new Literal(i), elemValueRef).genCode(ctx);
+        notNullAction.apply(new Reference(i), elemValueRef).genCode(ctx);
     if (nullAction == null) {
       String code =
           StringUtils.format(
@@ -127,7 +127,7 @@ public class ArrayDataForEach implements Expression {
               CodeGenerator.alignIndent(notNullElemExprCode.code(), 8));
       codeBuilder.append(code);
     } else {
-      Code.ExprCode nullExprCode = nullAction.apply(new Literal(i)).genCode(ctx);
+      Code.ExprCode nullExprCode = nullAction.apply(new Reference(i)).genCode(ctx);
       String code =
           StringUtils.format(
               ""

--- a/java/fury-format/src/main/java/io/fury/format/encoder/RowEncoderBuilder.java
+++ b/java/fury-format/src/main/java/io/fury/format/encoder/RowEncoderBuilder.java
@@ -27,6 +27,7 @@ import io.fury.builder.CodecBuilder;
 import io.fury.codegen.CodeGenerator;
 import io.fury.codegen.CodegenContext;
 import io.fury.codegen.Expression;
+import io.fury.codegen.Expression.Literal;
 import io.fury.codegen.Expression.Reference;
 import io.fury.codegen.ExpressionUtils;
 import io.fury.format.row.ArrayData;
@@ -81,12 +82,12 @@ public class RowEncoderBuilder extends BaseBinaryEncoderBuilder {
     ctx.reserveName(BEAN_CLASS_NAME);
     Expression clsExpr;
     if (Modifier.isPublic(beanClass.getModifiers())) {
-      clsExpr = new Expression.Literal(ctx.type(beanClass) + ".class");
+      clsExpr = Literal.ofClass(beanClass);
     } else {
       // non-public class is not accessible in other class.
       clsExpr =
           new Expression.StaticInvoke(
-              Class.class, "forName", CLASS_TYPE, false, Expression.Literal.ofClass(beanClass));
+              Class.class, "forName", CLASS_TYPE, false, Literal.ofClass(beanClass));
     }
     ctx.addField(Class.class, "beanClass", clsExpr);
     ctx.addImports(Field.class, Schema.class);
@@ -165,7 +166,7 @@ public class RowEncoderBuilder extends BaseBinaryEncoderBuilder {
       Preconditions.checkNotNull(d);
       TypeToken<?> fieldType = d.getTypeToken();
       Expression fieldValue = getFieldValue(bean, d);
-      Expression.Literal ordinal = new Expression.Literal(String.valueOf(i));
+      Literal ordinal = Literal.ofInt(i);
       Expression.StaticInvoke field =
           new Expression.StaticInvoke(
               DataTypes.class, "fieldOfSchema", ARROW_FIELD_TYPE, false, schemaExpr, ordinal);
@@ -191,7 +192,7 @@ public class RowEncoderBuilder extends BaseBinaryEncoderBuilder {
     expressions.add(bean);
     // schema field's name must correspond to descriptor's name.
     for (int i = 0; i < numFields; i++) {
-      Expression.Literal ordinal = new Expression.Literal(String.valueOf(i));
+      Literal ordinal = Literal.ofInt(i);
       Descriptor d = getDescriptorByFieldName(schema.getFields().get(i).getName());
       TypeToken<?> fieldType = d.getTypeToken();
       Expression.Invoke isNullAt =

--- a/java/fury-test-core/src/main/java/io/fury/test/bean/AccessBeans.java
+++ b/java/fury-test-core/src/main/java/io/fury/test/bean/AccessBeans.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2023 The Fury Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package io.fury.test.bean;
 
 import lombok.AllArgsConstructor;

--- a/java/fury-test-core/src/main/java/io/fury/test/bean/AccessBeans.java
+++ b/java/fury-test-core/src/main/java/io/fury/test/bean/AccessBeans.java
@@ -1,0 +1,63 @@
+package io.fury.test.bean;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+public class AccessBeans {
+  @Data
+  @AllArgsConstructor
+  private static class PrivateClass {
+    public int f1;
+    int f2;
+    private int f3;
+  }
+
+  @Data
+  @AllArgsConstructor
+  private static final class FinalPrivateClass {
+    public int f1;
+    int f2;
+    private int f3;
+  }
+
+  @Data
+  @AllArgsConstructor
+  static class DefaultLevelClass {
+    public int f1;
+    int f2;
+    private int f3;
+  }
+
+  @Data
+  @AllArgsConstructor
+  public static class PublicClass {
+    public int f1;
+    int f2;
+    private int f3;
+    private DefaultLevelClass f4;
+    private PrivateClass f5;
+    private FinalPrivateClass f6;
+  }
+
+  public static PrivateClass createPrivateClassObject() {
+    return new PrivateClass(1, 2, 3);
+  }
+
+  public static FinalPrivateClass createPrivateFinalClassObject() {
+    return new FinalPrivateClass(1, 2, 3);
+  }
+
+  public static DefaultLevelClass createDefaultLevelClassObject() {
+    return new DefaultLevelClass(4, 5, 6);
+  }
+
+  public static PublicClass createPublicClassObject() {
+    return new PublicClass(
+        1,
+        2,
+        3,
+        createDefaultLevelClassObject(),
+        createPrivateClassObject(),
+        createPrivateFinalClassObject());
+  }
+}

